firewall: T7318: Add support for nftables map

l0crian1 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change summary
<!--- Provide a general summary of your changes in the Title above -->
This change will add the ability to configure nft maps in VyOS. This can radically improve the speed of rule processing in deployments with a large number of firewall rules, since rules are not evaluated sequentially, but against a hashmap.

### Syntax:
#### Add parameters
```
set firewall group ipv4-map test parameters source-address	
set firewall group ipv4-map test parameters destination-address	
set firewall group ipv4-map test parameters destination-port
set firewall group ipv4-map test parameters protocol-tcp
```
#### Add rules
```
set firewall group ipv4-map test rule 10 action 'accept'
set firewall group ipv4-map test rule 10 destination address '10.0.5.0/24'
set firewall group ipv4-map test rule 10 destination port '179'
set firewall group ipv4-map test rule 10 source address '0.0.0.0/0'

set firewall group ipv4-map test rule 20 action 'drop'
set firewall group ipv4-map test rule 20 destination address '10.0.5.0/24'
set firewall group ipv4-map test rule 20 destination port '3389'
set firewall group ipv4-map test rule 20 source address '0.0.0.0/0'
```
After committing these, the map will be applied. You can verify in nftables:
```
vyos@cli-dev# sudo nft list map vyos_filter MAP_test
table ip vyos_filter {
        map MAP_test {
                type ipv4_addr . ipv4_addr . inet_service . inet_proto : verdict
                flags interval
                counter
                elements = { 0.0.0.0/0 . 10.0.5.0/24 . 179 . tcp counter packets 0 bytes 0 : accept,
                             0.0.0.0/0 . 10.0.5.0/24 . 3389 . tcp counter packets 0 bytes 0 : drop }
        }
}
```
#### Apply map to a rule
```
set firewall ipv4 output filter rule 10 action 'apply-map'
set firewall ipv4 output filter rule 10 map 'test'
```
After applying the map, you can see the applied rule in nftables:
```
vyos@cli-dev# sudo nft list chain vyos_filter VYOS_OUTPUT_filter
table ip vyos_filter {
        chain VYOS_OUTPUT_filter {
                type filter hook output priority filter; policy accept;
                ip saddr . ip daddr . tcp dport . meta l4proto vmap @MAP_test comment "ipv4-OUT-filter-10"
                counter packets 50 bytes 8588 accept comment "OUT-filter default-action accept"
        }
}
```
Or with the normal `show firewall` commands:
```
vyos@cli-dev# run show firewall ipv4 output filter
Ruleset Information

---------------------------------
ipv4 Firewall "output filter"

Rule     Action     Protocol      Packets    Bytes  Conditions
-------  ---------  ----------  ---------  -------  -------------------------------------------------------------
10       apply-map  all                 0        0  ip saddr . ip daddr . tcp dport . meta l4proto vmap @MAP_test
default  accept     all               501    74416
```
Packet/Byte counts can be viewed directly in the map:
```
vyos@cli-dev# sudo nft list map vyos_filter MAP_test
table ip vyos_filter {
        map MAP_test {
                type ipv4_addr . ipv4_addr . inet_service . inet_proto : verdict
                flags interval
                counter
                elements = { 0.0.0.0/0 . 10.0.5.0/24 . 179 . tcp counter packets 22 bytes 1320 : accept,
                             0.0.0.0/0 . 10.0.5.0/24 . 3389 . tcp counter packets 0 bytes 0 : drop }
        }
}
```
The map can be searched for matched traffic with this (an op mode command will be added after this is merged):
```
vyos@cli-dev# sudo nft get element ip vyos_filter MAP_test { 10.0.95.207 . 10.0.5.1 . 179 . tcp }
table ip vyos_filter {
        map MAP_test {
                type ipv4_addr . ipv4_addr . inet_service . inet_proto : verdict
                flags interval
                elements = { 0.0.0.0/0 . 10.0.5.0/24 . 179 . tcp counter packets 32 bytes 1920 : accept }
        }
}
```
#### Possible parameters:
```
   destination-address  Allow destination address as parameter
   destination-port     Allow destination port as parameter
   icmp-code            Allow ICMP code as parameter
   icmp-type            Allow ICMP type as parameter
   inbound-interface    Allow inbound interface as parameter
   outbound-interface   Allow outbound interface as parameter
   protocol-icmp        Define protocol parameter as ICMP
   protocol-tcp         Define protocol parameter as TCP
   protocol-udp         Define protocol parameter as UDP
   source-address       Allow source address as parameter
   source-port          Allow source port as parameter
```
## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->
https://vyos.dev/T7318

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## How to test / Smoketest result
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```

Or provide the output of the smoketest

```
$ /usr/libexec/vyos/tests/smoke/cli/test_xxx_feature.py
test_01_simple_options (__main__.TestFeature.test_01_simple_options) ... ok
```
-->
```
test_ipv4_map (__main__.TestFirewall.test_ipv4_map) ... ok
test_ipv6_map (__main__.TestFirewall.test_ipv6_map) ... ok
```

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [x] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

l0crian1 committed:
firewall: T7318: Add support for nftables map

Usage:

set firewall group ipv4-map test parameters source-address
set firewall group ipv4-map test parameters destination-address
set firewall group ipv4-map test parameters destination-port
set firewall group ipv4-map test parameters protocol-tcp

set firewall group ipv4-map test rule 10 action 'accept'
set firewall group ipv4-map test rule 10 destination address '10.0.5.0/24'
set firewall group ipv4-map test rule 10 destination port '179'
set firewall group ipv4-map test rule 10 source address '0.0.0.0/0'

set firewall ipv4 output filter rule 10 action 'apply-map'
set firewall ipv4 output filter rule 10 map 'test'

l0crian1 committed:
firewall: T7318: Add support for nftables map

Usage:

set firewall group ipv4-map test parameters source-address
set firewall group ipv4-map test parameters destination-address
set firewall group ipv4-map test parameters destination-port
set firewall group ipv4-map test parameters protocol-tcp

set firewall group ipv4-map test rule 10 action 'accept'
set firewall group ipv4-map test rule 10 destination address '10.0.5.0/24'
set firewall group ipv4-map test rule 10 destination port '179'
set firewall group ipv4-map test rule 10 source address '0.0.0.0/0'

sever-sever commented:
CI does not like your changes:
```
DEBUG - Running Testcase: /usr/libexec/vyos/tests/smoke/cli/test_policy_route.py
DEBUG - test_pbr_group (__main__.TestPolicyRoute.test_pbr_group) ... ERROR
DEBUG - test_pbr_group (__main__.TestPolicyRoute.test_pbr_group) ... FAIL
DEBUG - test_pbr_mark (__main__.TestPolicyRoute.test_pbr_mark) ... ERROR
DEBUG - test_pbr_mark (__main__.TestPolicyRoute.test_pbr_mark) ... FAIL
DEBUG - test_pbr_mark_connection (__main__.TestPolicyRoute.test_pbr_mark_connection) ... ERROR
DEBUG - test_pbr_mark_connection (__main__.TestPolicyRoute.test_pbr_mark_connection) ... FAIL
DEBUG - test_pbr_matching_criteria (__main__.TestPolicyRoute.test_pbr_matching_criteria) ... ERROR
DEBUG - test_pbr_matching_criteria (__main__.TestPolicyRoute.test_pbr_matching_criteria) ... FAIL
DEBUG - test_pbr_table (__main__.TestPolicyRoute.test_pbr_table) ... ERROR
DEBUG - test_pbr_table (__main__.TestPolicyRoute.test_pbr_table) ... FAIL
DEBUG - test_pbr_vrf (__main__.TestPolicyRoute.test_pbr_vrf) ... ERROR
DEBUG - test_pbr_vrf (__main__.TestPolicyRoute.test_pbr_vrf) ... FAIL
```

l0crian1 committed:
firewall: T7318: Add support for nftables map

Fixed unsafe key call in dict: line 591 /python/vyos/firewall.py

Usage:

set firewall group ipv4-map test parameters source-address
set firewall group ipv4-map test parameters destination-address
set firewall group ipv4-map test parameters destination-port
set firewall group ipv4-map test parameters protocol-tcp

set firewall group ipv4-map test rule 10 action 'accept'
set firewall group ipv4-map test rule 10 destination address '10.0.5.0/24'
set firewall group ipv4-map test rule 10 destination port '179'
set firewall group ipv4-map test rule 10 source address '0.0.0.0/0'

set firewall ipv4 output filter rule 10 action 'apply-map'
set firewall ipv4 output filter rule 10 map 'test'

l0crian1 commented:
> CI does not like your changes:

@sever-sever 

Thanks, I didn't notice. I had a key call that was failing because I didn't realize policy routes called the parse_rule() function. Added safer key calls there, and a couple other parts. Tests should complete now.

```
test_pbr_group (__main__.TestPolicyRoute.test_pbr_group) ... ok
test_pbr_mark (__main__.TestPolicyRoute.test_pbr_mark) ... ok
test_pbr_mark_connection (__main__.TestPolicyRoute.test_pbr_mark_connection) ... ok
test_pbr_matching_criteria (__main__.TestPolicyRoute.test_pbr_matching_criteria) ... ok
test_pbr_table (__main__.TestPolicyRoute.test_pbr_table) ... ok
test_pbr_vrf (__main__.TestPolicyRoute.test_pbr_vrf) ... ok

----------------------------------------------------------------------
Ran 6 tests in 60.268s

OK
```

sarthurdev reviewed (commented):
Thanks for the PR. I understand the reasoning for this PR and not directly opposed to it be merged, when used - the performance gains should be measurable. Though at the same time I think a better solution is for an overall optimization to how firewall rules are processed and loaded.

We've previously discussed use of `nft --optimize` argument when loading rulesets (iirc there were bugs at the time that prevented its use)

I'll make a note to discuss this at next maintainer meeting and see what others think.

l0crian1 commented:
> Though at the same time I think a better solution is for an overall optimization to how firewall rules are processed and loaded.

Thanks for taking a look. I'll give you a little bit of why I saw the need for it in VyOS.

I targeted this at solutions for organizations that have very explicit firewall rules. Where every address, port, and protocol are allowed through a Port Open Request (POR). I've had customers with 20k rules in their firewall in this scheme. For that customer, they could optimistically eliminate 5-10% (probably less) of their rules through collapsing redundant policies, which would still leave 18-19k rules to traverse sequentially. 

In cases like that, being able to do a single lookup against a hash rather than sequential lookups would be crucial. Not just for traffic destined to be permitted, but also for traffic destined for a default deny, which would eat a lot of CPU cycles to get to the end of the ruleset.

`nft` optimizes on the key fields though, so it could reduce the rules by quite a bit more than just scrubbing overlapping port/ip ranges. If this were to be merged, one of the next things I wanted to add was a `import-from-csv` option, that would take a CSV matching the fields in the map, and then loading that. That would also eliminate the long commit/boot times of VyOS with large configs.

zdc commented on python/vyos/firewall.py:
```diff
@@ -522,6 +522,49 @@ def parse_rule(rule_conf, hook, fw_name, rule_id, ip_name):
+            map_expr.append(f'iifname')
+        # Add outbound interface if specified
+        if 'param_outbound_interface' in rule_conf:
+            map_expr.append(f'oifname')
```
Do not use f-strings without variables.

l0crian1 replied:
Done!

zdc commented on src/conf_mode/firewall.py:
```diff
@@ -157,6 +157,48 @@ def get_config(config=None):
+    # /python/vyos/firewall.py. This stores those values in the rule under the param_ prefix
+    if firewall.get('group', {}).get('ipv4_map', {}) or firewall.get('group', {}).get('ipv6_map', {}):
+        for family in ['ipv4', 'ipv6']:
+            if family in firewall:
+                for chain in ['name','forward','input','output']:
+                    if chain in firewall[family]:
+                        for priority, priority_conf in firewall[family][chain].items():
+                            if 'rule' in priority_conf:
+                                for rule_id, rule_conf in priority_conf['rule'].items():
+                                    if 'map' in rule_conf:
+                                        map_name = rule_conf['map']
+                                        map_family = f'{family}_map'
+
+                                        if firewall.get('group', {}).get(map_family, {}).get(map_name, {}).get('parameters'):
+                                            param_path = firewall['group'][map_family][map_name]['parameters']
+                                            rule_path = firewall[family][chain][priority]['rule'][rule_id]
+
+                                            if 'protocol_tcp' in param_path:
+                                                rule_path['param_protocol'] = 'tcp'
+                                            if 'protocol_udp' in param_path:
+                                              rule_path['param_protocol'] = 'udp'
+                                            if 'protocol_icmp' in param_path:
+                                                rule_path['param_protocol'] = 'icmp'
+                                                if 'icmp_type' in param_path:
+                                                    rule_path['param_icmp_type'] = {}
+                                                if 'icmp_code' in param_path:
+                                                    rule_path['param_icmp_code'] = {}
+                                            if 'source_address' in param_path:
+                                                rule_path['param_source_address'] = {}
+                                            if 'destination_address' in param_path:
+                                                rule_path['param_destination_address'] = {}
+                                            if 'source_port' in param_path:
+                                                rule_path['param_source_port'] = {}
+                                            if 'destination_port' in param_path:
+                                                rule_path['param_destination_port'] = {}
+                                            if 'inbound_interface' in param_path:
+                                                rule_path['param_inbound_interface'] = {}
+                                            if 'outbound_interface' in param_path:
+                                                rule_path['param_outbound_interface'] = {}
+
```
I'm not entirely sure why this is required, but the code itself can be optimized for better clarity and some performance improvement as well:

```
if firewall.get("group", {}).keys() & {"ipv4_map", "ipv6_map"}:
    _PARAM_MAP = {
        "protocol_tcp": ("param_protocol", "tcp"),
        "protocol_udp": ("param_protocol", "udp"),
        "protocol_icmp": ("param_protocol", "icmp"),
        "icmp_type": ("param_icmp_type", {}),
        "icmp_code": ("param_icmp_code", {}),
        "source_address": ("param_source_address", {}),
        "destination_address": ("param_destination_address", {}),
        "source_port": ("param_source_port", {}),
        "destination_port": ("param_destination_port", {}),
        "inbound_interface": ("param_inbound_interface", {}),
        "outbound_interface": ("param_outbound_interface", {}),
    }
    for family in ("ipv4", "ipv6"):
        family_conf = firewall.get(family, {})
        group_map = firewall.get("group", {}).get(f"{family}_map", {})

        for chain_conf in (
            family_conf.get(c, {}) for c in ("name", "forward", "input", "output")
        ):
            for priority_conf in chain_conf.values():
                rule_dict = priority_conf.get("rule", {})
                for rule_conf in rule_dict.values():
                    params = group_map.get(rule_conf.get("map"), {}).get("parameters", ())
                    if not params:
                        continue

                    # only loop over params that actually appear in both param‐map and the config
                    for key in set(params) & _PARAM_MAP.keys():
                        dest_field, value = _PARAM_MAP[key]
                        rule_conf[dest_field] = value
```

l0crian1 replied:
This is much cleaner, thanks.

This is needed because by the time the individual rule string for the chain is being created by `/python/vyos/firewall.py`, I can't get the parameters values with `get_config_dict()` since it seems to be spawning a different session of Config(), which returns as empty unless the map was commited before trying to commit the rule to a chain. So I store the values in the original rule so I can call them to create something like: ` ip saddr . ip daddr . tcp dport . meta l4proto vmap @MAP_test`.

l0crian1 replied:
Done!

zdc replied:
@jestabro could you please check this? I feel like something might be wrong here - either in our code or in the new code...

> This is needed because by the time the individual rule string for the chain is being created by `/python/vyos/firewall.py`, I can't get the parameters values with `get_config_dict()` since it seems to be spawning a different session of Config(), which returns as empty unless the map was commited before trying to commit the rule to a chain.

zdc commented on src/conf_mode/firewall.py:
```diff
@@ -198,14 +240,59 @@ def verify_rule(firewall, family, hook, priority, rule_id, rule_conf):
+        if not firewall.get('group', {}).get(f'{family}_map', {}).get(rule_conf['map']):
+            raise ConfigError(f'{family}-map "{rule_conf["map"]}" does not exist')
+        else:
+            map_conf = firewall.get('group', {}).get(f'{family}_map', {}).get(rule_conf['map'])
```
Maybe reading it once and reusing it would be shorter here?

```
map_conf = firewall.get("group", {}).get(f"{family}_map", {}).get(rule_conf["map"])
if not map_conf:
    raise ConfigError(f'{family}-map "{rule_conf["map"]}" does not exist')
```

l0crian1 replied:
Done!

zdc commented on src/conf_mode/firewall.py:
```diff
@@ -423,6 +510,106 @@ def verify_hardware_offload(ifname):
+                if len(map_conf['parameters']) > 5:
+                    raise ConfigError(f'{map_family} "{map_name}" cannot have more than 5 parameters defined')
+
+                param_list = [k for k in map_conf['parameters'] if "protocol" not in k]
```
To improve clarity and performance, consider avoiding repeated access to the same value:

```
map_conf.get('parameters', {})
map_conf['parameters']
```


It's better to read the value once (e.g., store it in a variable like `params`) and reuse it throughout the block.

l0crian1 replied:
Done!

zdc reviewed (changes requested):
The idea is great and definitely necessary.

I won't judge the CLI - there's no perfect one for everyone anyway, and someone from the CLI masters will probably review it as well.

As for the implementation, there are a few optimizations to consider.

Also, I tend not to consider the maps implementation complete without including it in NAT rules. Would you like to expand the PR to cover NAT as well, @l0crian1? 🙂

l0crian1 commented:
@zdc Thanks I'll start working on these.

> Also, I tend not to consider the maps implementation complete without including it in NAT rules. Would you like to expand the PR to cover NAT as well, @l0crian1? 🙂

I started with verdict maps because they were the most straightforward implementation. They're always `<some keys> : verdict`, whereas those used for NAT are a little more open with `<some keys> : value`. I think adding them as a separate PR would be better for me, but I'm fully fine with adding them. I also started with verdict maps because people will tend to have a lot more firewall rules than they will NAT statements.

That does present a question about syntax though. Would it make more sense to call them something like this?:
```
ipv4-verdict-map
ipv4-source-nat-map
ipv4-destination-nat-map

ipv6-verdict-map
ipv6-source-nat-map
ipv6-destination-nat-map
```

l0crian1 committed:
T7318: Add support for nftables map

Requested updates from zdc

zdc commented:
> I started with verdict maps because they were the most straightforward implementation. They're always `<some keys> : verdict`, whereas those used for NAT are a little more open with `<some keys> : value`. I think adding them as a separate PR would be better for me, but I'm fully fine with adding them. I also started with verdict maps because people will tend to have a lot more firewall rules than they will NAT statements.

I am OK with two PRs in general. My expectation is that both filter and NAT share the same part of the code and CLI. I prefer to see at least the CLI before we merge one of these parts. The problem we encounter is that when we implement half of a feature and then come back with the other half, it often turns out that the CLI must be changed to properly support both parts.

And this brings us to:

> That does present a question about syntax though. Would it make more sense to call them something like this?:
> 
> ```
> ipv4-verdict-map
> ipv4-source-nat-map
> ipv4-destination-nat-map
> 
> ipv6-verdict-map
> ipv6-source-nat-map
> ipv6-destination-nat-map
> ```

I would prefer to have two types: `verdict-map` and `map`, and validate their content for each firewall/NAT rule. The idea is that the same type of map can be used across various rule types (SNAT/DNAT), even outside of NAT entirely, for other features later.

l0crian1 commented:
> I am OK with two PRs in general. My expectation is that both filter and NAT share the same part of the code and CLI. I prefer to see at least the CLI before we merge one of these parts. The problem we encounter is that when we implement half of a feature and then come back with the other half, it often turns out that the CLI must be changed to properly support both parts.

Completely makes sense!

> I would prefer to have two types: `verdict-map` and `map`, and validate their content for each firewall/NAT rule. The idea is that the same type of map can be used across various rule types (SNAT/DNAT), even outside of NAT entirely, for other features later.

Maps have limited utility outside of verdicts and NAT. Even NAT is limiting because you can't do `flags interval`, all elements need to be a singleton; fine for DNAT, but if you want to use SNAT pools, it's more efficient to not use maps. The other uses are mostly mangling (e.g. setting mark/ct mark), and that seems fairly limiting as well (at least in the versions of nftables on VyOS).

My initial thoughts were having these commands for uniformity:
### Firewall:
```
ipv4-verdict-map
ipv6-verdict-map

ipv4-map
ipv6-map
```
**The only reason I even have the normal map in here is because you can `set` values in the FW rules in 1.5**
### NAT:
```
source ipv4-map
source ipv6-map

destination ipv4-map
destination ipv6-map
```
**Trying to use a single map that could be used for either SNAT/DNAT is problematic because the VyOS syntax doesn't match that. For instance, SNAT looks only for outbound-interface and DNAT looks for inbound-interface. It wouldn't really buy you anything, and would require additional validation/verification when there's no efficiency gains or reduced memory footprint of the maps.**
### Policy:
```
ipv4-map
ipv6-map
```

Because the syntax can vary widely between Firewall, NAT, and Policy, trying to make a more generic map that could be applied to FW, NAT, or Policy would not really work well without some departures in VyOS syntax and conventions. Keeping the map configs within each section makes all of the configuration familiar to the user.

zdc commented:
You are technically right. Let me explain some nuances to clarify the details we pursue.

> Maps have limited utility outside of verdicts and NAT
Right now - yes, because we do not support many nftables features in the CLI. But maps are a powerful tool, which can be expanded later to other features, like `fwd` for example (a pretty extreme case, but we do a lot of crazy stuff here - do not be surprised).

> NAT is limiting because you can't do flags interval, all elements need to be a singleton; fine for DNAT, but if you want to use SNAT pools, it's more efficient to not use maps

This is specifically one of those cases where we prefer to have maps instead of tons of rules. Sometimes we need to manage NAT dynamically, by adding hundreds of addresses, ports, etc., on the fly or generating them from a template (for automated deployments). NAT maps are quite helpful here.

> seems fairly limiting as well (at least in the versions of nftables on VyOS)

> Trying to use a single map that could be used for either SNAT/DNAT is problematic because the VyOS syntax doesn't match that

You are totally right here! There is a lot of historical legacy we carry in the CLI and firewall logic, but we already made a big jump forward in 1.4 and restructured the CLI to allow adding much more in the future. Here is the main point - many things are useless right now, but we should think years in advance to avoid changing the CLI and internal logic multiple times later.

> ```
> ipv4-*
> ipv6-*
> ```

The amazing part of maps is that they can include basically any type from nftables, and this gives them superpowers. So, I can technically agree with something that contains the IP protocol version, if we limit the usage of these maps to `ip` or `ip6` families only. But then, there is also `inet` and other not-so-popular families.

I want to have more flexibility in the future and mix everything, like:

```
type inet_service : verdict
type ifname . inet_service : verdict
type ifname . inet_service . mark : verdict
type ifname . inet_service . mark : ipv4_addr . inet_service
```

It is hard to imagine how the CLI can remain straightforward in such a case if we include types right into a CLI parent object.


> It wouldn't really buy you anything, and would require additional validation/verification when there's no efficiency gains or reduced memory footprint of the maps.

In this specific case, it is more about having a unified CLI and less about efficiency.

> syntax can vary widely between Firewall, NAT, and Policy

This is something we should address one day.

> Keeping the map configs within each section makes all of the configuration familiar to the user.

I would say, this is our pain. There are so many improvements that cannot be released just because users are scared of CLI changes! :-D

So, summarizing...

What you did in this PR is really great! Starting from verdicts is a reasonable step because they are easier from all sides. But what I would like to see before finalizing the verdict CLI is a simple maps CLI prototype as well - no code, just how you see this in the end, with as many details as you think are reasonable - unleash your fantasy! This can significantly change how verdict maps are integrated as well, in the end.

l0crian1 commented:
> > Maps have limited utility outside of verdicts and NAT

> Right now - yes, because we do not support many nftables features in the CLI. But maps are a powerful tool, which can be expanded later to other features, like `fwd` for example (a pretty extreme case, but we do a lot of crazy stuff here - do not be surprised).

I didn't just mean in the context of VyOS, I meant in general with nftables. You can match a lot of key fields, but the value field has limited, albeit powerful, uses. Primarily verdict, NAT, and marking. You can match a few more values in that field that can't actually be applied in a rule., For instance `ipv4_addr : dscp` is accepted in the map, but the functionality to use that in a rule doesn't work. The map even has a limit of 5 keys, so you can't get too crazy with matches. Obviously, with me submitting this PR, I'm not saying maps aren't useful, they're just not as flexible as they may seem on paper (in function if not conceptually).

> > NAT is limiting because you can't do flags interval, all elements need to be a singleton; fine for DNAT, but if you want to use SNAT pools, it's more efficient to not use maps
> 
> This is specifically one of those cases where we prefer to have maps instead of tons of rules. Sometimes we need to manage NAT dynamically, by adding hundreds of addresses, ports, etc., on the fly or generating them from a template (for automated deployments). NAT maps are quite helpful here.

Totally agree on how useful they can be for NAT. What I meant was wanting to SNAT an inside network of 10.0.0.0/24 to an outside network of say 203.0.113.0/29. You can't do `10.0.0.0/24 : 203.0.113.0/29` in a map. If the value field is an address, both sides must be a singleton, you can't use the interval, even though the map accepts it as a flag. It's very useful for DNAT since that syntax will usually be a single ip/port to another ip/port. So that is what I mean by limiting, but there's still a lot of value to allowing maps to be used with NAT.

> The amazing part of maps is that they can include basically any type from nftables, and this gives them superpowers. So, I can technically agree with something that contains the IP protocol version, if we limit the usage of these maps to `ip` or `ip6` families only. But then, there is also `inet` and other not-so-popular families.
> 
> I want to have more flexibility in the future and mix everything, like:
> 
> ```
> type inet_service : verdict
> type ifname . inet_service : verdict
> type ifname . inet_service . mark : verdict
> type ifname . inet_service . mark : ipv4_addr . inet_service
> ```
> 
> It is hard to imagine how the CLI can remain straightforward in such a case if we include types right into a CLI parent object.

This is already supported in the current syntax I have in this PR. Each key is effectively a parameter. So if `mark` would be a desired key, then it'd be as easy as just adding that as a parameter. The same would go for any additional parameter.

I don't view the syntax differences between firewall and NAT as a bad thing (outside of inconsistency). If someone is configuring NAT, then they don't need to even see options outside of NAT. Same for configuring firewall. This is just small sampling of things that only have relevance within certain sections. There's obviously a lot more.

| Keyword                                | Valid For              |
|----------------------------------------|-------------------------|
| action                                 | firewall and policy     |
| add-address-to-group                   | firewall                |
| conntrack-helper                       | firewall                |
| inbound-interface                      | firewall and DNAT       |
| jump-target                            | firewall                |
| load-balance                           | SNAT and DNAT           |
| offload-target                         | firewall                |
| outbound-interface                     | firewall and SNAT       |
| queue                                  | firewall                |
| queue-options                          | firewall                |
| set                                    | firewall and policy     |
| synproxy                               | firewall                |
| translation                            | SNAT and DNAT           |
| translation address masquerade         | SNAT                    |
| translation address redirect           | DNAT                    |

If you put everything in a single configurable map section (like `ipv4-map`), you'd need a massive validation effort to ensure everything works well. Then if someone did add a key to the maps section for say a firewall field, they'd have to remember (or even know) that they need to add validation checks for NAT and policy. Compare that to adding a key to a map that was just configurable under the firewall section, where it wouldn't impact the NAT and policy implementation.

There's already probably 100 missing validation checks in the firewall section (some that could actually cause inadvertently dropping/accepting all traffic), and I'd imagine that would become the case here as well if everything were to be dumped into a single map.

> > Keeping the map configs within each section makes all of the configuration familiar to the user.
> 
> I would say, this is our pain. There are so many improvements that cannot be released just because users are scared of CLI changes! :-D

I wasn't really referencing an aversion to CLI changes; if they make sense, I say change them. I mean if you're configuring a map for DNAT, then seeing 50+ additional options for things outside of DNAT doesn't help anything. It's just clutter.

For me, I think having a specific map config per section is better. A single map config is better conceptually, but I don't think it is functionally or practically.

l0crian1 commented:
I'm going to close this for now. I'll revisit if I find time to implement it as desired.
